stdlib/public/runtime/DynamicCast.cpp

```diff
@@ -425,6 +425,30 @@ tryCastUnwrappingObjCSwiftValueSource(
       destFailureType, srcFailureType,
       /*takeOnSuccess=*/ false, mayDeferChecks);
 }
+#else
+static DynamicCastResult
+tryCastUnwrappingSwiftValueSource(
+  OpaqueValue *destLocation, const Metadata *destType,
+  OpaqueValue *srcValue, const Metadata *srcType,
+  const Metadata *&destFailureType, const Metadata *&srcFailureType,
+  bool takeOnSuccess, bool mayDeferChecks)
+{
+    const Metadata *srcInnerType;
+    const OpaqueValue *srcInnerValue;
+    if (swift_unboxFromSwiftValueWithType(srcValue,
+       reinterpret_cast<OpaqueValue *>(&srcInnerValue), &srcInnerType)) {
+      return DynamicCastResult::SuccessViaCopy;
+    }
+
+    // Note: We never `take` the contents from a SwiftValue box as
+    // it might have other references.  Instead, let our caller
+    // destroy the reference if necessary.
+    return tryCast(
+      destLocation, destType,
+      const_cast<OpaqueValue *>(srcInnerValue), srcInnerType,
+      destFailureType, srcFailureType,
+      /*takeOnSuccess=*/ false, mayDeferChecks);
```

Comment:
I don't think this is right. If `swift_unboxFromSwiftValueWithType` succeeds, then we return success but nothing places the value into `destLocation`. If it fails, then we use the uninitialized `srcInnerValue` with `tryCast`. The third parameter is the expected type of the inner value, so `&srcInnerType` isn't going to do what we want. I think the right approach here would be to either grab `value` directly from the `__SwiftValue` object by offset, or call `swift_unboxFromSwiftValueWithType` with `Any` as the destination type, then continue casting with the output value (and, presumably, fail loudly if that unboxing somehow fails).

Reply:
Yes, I did a bad copy/paste job here.  Looking at it more carefully, it looks like `swift_unboxFromSwiftValueWithType` is really just a casting operation itself, so invoking it to recursively perform the cast is probably the easiest approach.  Let's see what CI thinks of that.

Reply:
When I wrote my first comment, I somehow didn't realize that all the `as? T` expressions in the implementation of `swift_unboxFromSwiftValueWithType` will actually result in full-fledged casting, and I was thinking that it would have to be an exact match. This new version is nice!